### PR TITLE
BP-834: Update to Angular 16

### PR DESCRIPTION
**Pull request by msljivic**

### CODE REVIEW

File: package.json [hunk 1]
```diff
@@ -46,25 +46,25 @@
...
+    "@angular-eslint/eslint-plugin-template": "16.3.1",
+    "@angular-eslint/schematics": "16.3.1",
+    "@angular-eslint/template-parser": "16.3.1",
+    "@angular-material-components/datetime-picker": "16.0.1",
+    "@angular/cli": "^16.2.12",
+    "@angular/compiler-cli": "^16.2.12",
+    "@angular/material-date-fns-adapter": "16.2.14",
     "@ckeditor/ckeditor5-angular": "^5.2.0",
     "@ckeditor/ckeditor5-build-classic": "^37.0.0",
     "@enigmatry/scss-foundation": "15.0.0-preview.5",
```
Comment (msljivic) [inline package.json:64]: Not sure why are we installing @enigmatry/scss-foundation and @enigmatry/stylelint-config as devDependencies

File: .eslintrc.json [hunk 2]
```diff
@@ -16,8 +16,9 @@
         "createDefaultProgram": true
       },
       "extends": [
-        "plugin:@angular-eslint/ng-cli-compat",
-        "plugin:@angular-eslint/ng-cli-compat--formatting-add-on",
+        "eslint:recommended",
+        "plugin:@typescript-eslint/recommended",
+        "plugin:@angular-eslint/recommended",
         "plugin:@angular-eslint/template/process-inline-templates"
```
Comment (msljivic) [inline .eslintrc.json:22]: Configure eslint to use recommended rules for demo and libs.

File: .eslintrc.json [hunk 3]
```diff
@@ -37,8 +38,14 @@
             "style": "camelCase"
           }
         ],
-        "no-underscore-dangle": "off",
-        "prefer-arrow/prefer-arrow-functions": "off"
+        "prefer-arrow/prefer-arrow-functions": "off",
+        "@typescript-eslint/no-unused-vars": [
+          "error",
+          {
+            "argsIgnorePattern": "^_",
```
Comment (msljivic) [inline .eslintrc.json:45]: Extend no-unused-vars with ignore patterns.

File: .eslintrc.json (hunk 2 above)
Comment (zsrdjan) [reply .eslintrc.json:22]: I usually generate new angular project, see how things are setup there, and copy the setup with as minimal adjustments needed to make future maintenance simpler.

File: apps/demo-app/src/app/examples/search-filter/search-filter/search-filter-example.component.ts [hunk 4]
```diff
@@ -36,7 +36,7 @@ export class SearchFilterExampleComponent {
     this.users = this._usersService.getUsers(searchParams);
   }
 
-  private createSearchFilters(): SearchFilterBase<any>[] {
+  private createSearchFilters(): SearchFilterBase<unknown>[] {
```
Comment (zsrdjan) [inline apps/demo-app/src/app/examples/search-filter/search-filter/search-filter-example.component.ts:39]: yet another type in typescript, bah

File: apps/demo-app/src/app/examples/dialog/error/error-example.component.ts [hunk 5]
```diff
@@ -18,6 +18,6 @@ export class ErrorDialogExampleComponent {
           errors: err
         })
       });
-    ;
-  };
+    
+  }
```
Comment (zsrdjan) [inline apps/demo-app/src/app/examples/dialog/error/error-example.component.ts:22]: Is this new typescript feature?

File: apps/demo-app/src/app/examples/search-filter/search-filter/users.ts [hunk 6]
```diff
@@ -32,7 +32,7 @@ enum Country {
     southAfrica = 'South Africa'
 }
 
-class User {
+interface User {
```
Comment (zsrdjan) [inline apps/demo-app/src/app/examples/search-filter/search-filter/users.ts:35]: Is this changed related to the angular upgrade, and if yes, how?

File: libs/entry-components/.eslintrc.json [hunk 7]
```diff
@@ -31,7 +31,8 @@
             "prefix": "entry",
             "style": "kebab-case"
           }
-        ]
+        ],
+        "@typescript-eslint/no-explicit-any": "off"
```
Comment (zsrdjan) [inline libs/entry-components/.eslintrc.json:35]: Why did you turn off this one?

File: apps/demo-app/src/app/examples/dialog/error/error-example.component.ts (hunk 5 above)
Comment (msljivic) [reply apps/demo-app/src/app/examples/dialog/error/error-example.component.ts:22]: lint --fix -> not needed semicolons removed

File: apps/demo-app/src/app/examples/search-filter/search-filter/search-filter-example.component.ts (hunk 4 above)
Comment (msljivic) [reply apps/demo-app/src/app/examples/search-filter/search-filter/search-filter-example.component.ts:39]: unknown is consider a "type-safe" alternative to any

https://www.totaltypescript.com/concepts/object-is-of-type-unknown

With any, you can access any property on it without narrowing it.
This is not possible with unknown

Comment (typhoon41) [reply apps/demo-app/src/app/examples/search-filter/search-filter/search-filter-example.component.ts:39]: We should forbid usage of any. No point of using TypeScript (typed JS) if you prefer using any. You can sort out 99.99% of cases with unknowns. It's been there for years. You can still use nice features of JS such as accessing properties in objects in a dictionary way with use of Record<> and other rich type features added in versions 4.x

File: apps/demo-app/src/app/examples/search-filter/search-filter/users.ts (hunk 6 above)
Comment (msljivic) [reply apps/demo-app/src/app/examples/search-filter/search-filter/users.ts:35]: not directly related to angular update

Comment (msljivic) [reply apps/demo-app/src/app/examples/search-filter/search-filter/users.ts:35]: removing this nonsense below 
```diff
-
-    constructor(data?: Partial<User>) {
-        if (data) {
-            for (const property in data) {
-                if (data.hasOwnProperty(property)) { (this)[property] = (data)[property]; }
-            }
-        }
-    }
```

File: libs/entry-components/.eslintrc.json (hunk 7 above)
Comment (msljivic) [reply libs/entry-components/.eslintrc.json:35]: We are using any in more then few places in the library.
Maybe there is good justification for most of them, we can investigate this in a separate issue if you agree.

e.g.
https://github.com/enigmatry/entry-angular-building-blocks/blob/adfea80b1557f0be59bedaa09b392a92d0eac41f/libs/entry-components/dialog/dialogs/entry-dialog.component.ts#L38

https://github.com/enigmatry/entry-angular-building-blocks/blob/adfea80b1557f0be59bedaa09b392a92d0eac41f/libs/entry-components/dialog/entry-dialog.service.ts#L69

**Review by zsrdjan — APPROVED** [review]
